File: catalog.py
#!/usr/bin/env python3
"""Site catalog: JSON file or WAF_SITES_JSON. No production hosts in source."""
from __future__ import annotations

import json
import os
from pathlib import Path

import netguard

CONTROL_DIR = Path(os.environ.get("WAF_CONTROL", "/var/lib/waf-control"))


def _parse(raw: str) -> dict:
    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        return {}
    return data if isinstance(data, dict) else {}


def _clean_host(value: str) -> str | None:
    try:
        return netguard.sanitize_hostname(str(value))
    except ValueError:
        return None


def load_catalog() -> dict:
    env = os.environ.get("WAF_SITES_JSON", "").strip()
    if env:
        data = _parse(env)
    else:
        path = Path(os.environ.get("WAF_SITES_FILE", str(CONTROL_DIR / "sites.json")))
        if path.is_file():
            data = _parse(path.read_text(encoding="utf-8", errors="replace"))
        else:
            data = {}
    sites_raw = data.get("sites") if isinstance(data.get("sites"), dict) else {}
    oos_raw = data.get("out_of_scope") if isinstance(data.get("out_of_scope"), dict) else {}
    sites = {}
    for host, meta in sites_raw.items():
        clean = _clean_host(host)
        if clean and isinstance(meta, dict):
            sites[clean] = meta
    oos = {}
    for host, why in oos_raw.items():
        clean = _clean_host(host)
        if clean:
            oos[clean] = why
    hosts = data.get("hosts")
    if not isinstance(hosts, list):
        hosts = [h for h, meta in sites.items() if meta.get("in_scope", True)]
    cleaned_hosts = []
    for h in hosts:
        clean = _clean_host(h)
        if clean:
            cleaned_hosts.append(clean)
    return {"sites": sites, "out_of_scope": oos, "hosts": cleaned_hosts}


def in_scope_hosts() -> list[str]:
    env = os.environ.get("WAF_HOSTS", "").strip()
    if env:
        out = []
        for h in env.split(","):
            clean = _clean_host(h)
            if clean:
                out.append(clean)
        return out
    return load_catalog()["hosts"]

File: rust/src/catalog.rs
//! Site catalog (port of catalog.py).

use std::path::PathBuf;

use serde_json::{json, Map, Value};

use crate::netguard;

fn control_dir() -> PathBuf {
    PathBuf::from(std::env::var("WAF_CONTROL").unwrap_or_else(|_| "/var/lib/waf-control".into()))
}

fn parse(raw: &str) -> Map<String, Value> {
    serde_json::from_str(raw)
        .ok()
        .and_then(|v: Value| v.as_object().cloned())
        .unwrap_or_default()
}

fn clean_host(value: &Value) -> Option<String> {
    netguard::sanitize_hostname(value.as_str().unwrap_or("")).ok()
}

pub fn load_catalog() -> Value {
    let data = if let Ok(env) = std::env::var("WAF_SITES_JSON") {
        if !env.trim().is_empty() {
            parse(&env)
        } else {
            Map::new()
        }
    } else {
        let path = std::env::var("WAF_SITES_FILE")
            .map(PathBuf::from)
            .unwrap_or_else(|_| control_dir().join("sites.json"));
        if path.is_file() {
            parse(&std::fs::read_to_string(path).unwrap_or_default())
        } else {
            Map::new()
        }
    };
    let sites_raw = data
        .get("sites")
        .and_then(|v| v.as_object())
        .cloned()
        .unwrap_or_default();
    let oos_raw = data
        .get("out_of_scope")
        .and_then(|v| v.as_object())
        .cloned()
        .unwrap_or_default();
    let mut sites = Map::new();
    for (host, meta) in sites_raw {
        if let (Some(clean), Some(obj)) =
            (clean_host(&Value::String(host.clone())), meta.as_object())
        {
            sites.insert(clean, Value::Object(obj.clone()));
        }
    }
    let mut oos = Map::new();
    for (host, why) in oos_raw {
        if let Some(clean) = clean_host(&Value::String(host)) {
            oos.insert(clean, why);
        }
    }
    let hosts = if let Some(list) = data.get("hosts").and_then(|v| v.as_array()) {
        list.iter()
            .filter_map(|h| clean_host(h))
            .collect::<Vec<_>>()
    } else {
        sites
            .iter()
            .filter(|(_, meta)| {
                meta.get("in_scope")
                    .and_then(|v| v.as_bool())
                    .unwrap_or(true)
            })
            .map(|(h, _)| h.clone())
            .collect()
    };
    json!({ "sites": sites, "out_of_scope": oos, "hosts": hosts })
}

pub fn in_scope_hosts() -> Vec<String> {
    if let Ok(env) = std::env::var("WAF_HOSTS") {
        if !env.trim().is_empty() {
            return env
                .split(',')
                .filter_map(|h| clean_host(&Value::String(h.to_string())))
                .collect();
        }
    }
    load_catalog()
        .get("hosts")
        .and_then(|v| v.as_array())
        .map(|a| {
            a.iter()
                .filter_map(|h| h.as_str().map(String::from))
                .collect()
        })
        .unwrap_or_default()
}
